### ISSUE TITLE
onDidChangeTextDocument called when closing an unsaved document.

### ISSUE THREAD

**Issue by Friday32** [microsoft/vscode#81902]
onDidChangeTextDocument() is fired when you make changes in a document.  The event is also fired when you do not save the document and close it even through you made no further changes.   There are a number of issues with this behavior.

1.  The event for the same change should not fire twice. When closing the document, it has not changed since the last event call. 
2. The range included with the TextDocumentChangeEvent parameter is incorrect.  For example if I type a single character on line 12, the change range spans from line 12, position 0 to line 13, position 0.  
3,. There is no way to determine why the method is called. (See issue #36939).   

My extension may modify the document in response to user input. It depends on the region where the document is modified.  The second problem triggers my extension to modify the unsaved document when the user closes it.  This is turn triggers the document to be reopened.

**Comment by Friday32** [comment]
[example.zip](https://github.com/microsoft/vscode/files/3687407/example.zip)

I created a stripped down example to implement a handler for onDidChangeTextDocument.  

- Open the example in VSCode.
- Run npm install.
- Run "Launch Main" from launch.json. This will open another instance of VSCode for debugging.
- From the new instance, open a *.c text file.  
- Type anywhere in the file.  You should see some console output displaying the range of the first contentChange member.
- Close the document by by pressing the X icon on the associated tab.
- You should see the event onDidChangeTextDocument called again with the problem I described.

I'm using VSCode version 1.38.1

**Comment by jrieken** [comment]
> You should see the event onDidChangeTextDocument called again with the problem I described.

I don't see anything unusual here. The document is closed and no further event happens. Please update you sample and update your expected/actual behaviour steps. Tho, you sample is very misleading because you simply log that a change has happened. The `onDidChangeTextDocument` is fired for *all* documents and since you aren't logging the document uri that has changed it can be anything

**Comment by Friday32** [comment]
Only the Welcome document opens when I run the example. I close it and only open one document.  There are no other documents opened.   

All I do is:
- run example. A new instance of vscode opens with only welcome document opened.
- close welcome document.
- open C file.
- type in document.
- look at other instance of VSCode to see how many log statements are produced.
- close C file.
- observe a new, unexpected log statement has occurred.

I just reproduced this issue on another a computer downing the attached example. I removed the only extension I have installed, Microsoft cpptools.  I tried it again and observed the same issue. 

I was using a breakpoint to inspect other members of the parameter passed in by the event including the uri.  So, I am 100% sure that it came from the same document before submitting this bug. The only update I would make to the example is log the uri, but that's not going help much.  

 The problem occurs easily for me, so I figured you would have no trouble reproducing it.

**Comment by jrieken** [comment]
> So, I am 100% sure that it came from the same document before submittin

Just print the names please. document !== editor, e.g there are documents that you don't see as editor.

**Comment by Friday32** [comment]
Here's the document in json format that I receive after closing:

{"uri":{"$mid":1,"fsPath":"c:\\temp\\Untitled-1.c","_sep":1,"external":"file:///c%3A/temp/Untitled-1.c","path":"/c:/temp/Untitled-1.c","scheme":"file"},"fileName":"c:\\temp\\Untitled-1.c","isUntitled":false,"languageId":"c","version":4,"isClosed":false,"isDirty":true,"eol":2,"lineCount":2}

The last one from the last edit is:

{"uri":{"$mid":1,"fsPath":"c:\\temp\\Untitled-1.c","_sep":1,"external":"file:///c%3A/temp/Untitled-1.c","path":"/c:/temp/Untitled-1.c","scheme":"file"},"fileName":"c:\\temp\\Untitled-1.c","isUntitled":false,"languageId":"c","version":4,"isClosed":false,"isDirty":true,"eol":2,"lineCount":2}


I also added a check to verify that there is at least 1 content change. Sometimes, a call is made with no changes.  (Could be considered a bug.)

**Comment by jrieken** [comment]
> Sometimes, a call is made with no changes. (Could be considered a bug.)

Please read [this](https://github.com/microsoft/vscode/blob/d10c9f75163176df87817ef06750365cd76976be/src/vs/vscode.d.ts#L8134-L8136)

> Here's the document in json format that I receive after closing:

Any chance you have a formatter or anything else running? Please try without any other extension enabled

**Comment by Friday32** [comment]
[example.zip](https://github.com/microsoft/vscode/files/3691119/example.zip) (updated with minor changes)

I manually deleted all my extensions and still have the same issue.  If something else is running, I'm not sure what it could be.

**Comment by jrieken** [comment]
Sorry, I don't not see a change event after closing

**Comment by Friday32** [comment]
I had someone in my company's audit department try the example. He had no trouble reproducing the issue when running a debug session.  

I sent you a video of the issue last Friday via email.  Maybe that will help.

**Comment by jrieken** [comment]
Did you try this with the `--disable-extensions` flag?

**Comment by Friday32** [comment]
Yes, I tried it after you mentioned it.  It didn't make a difference.

Are you using the default  off setting for  "Files: Auto Save" under "Commonly Used"?

**Comment by jrieken** [comment]
YEah, but do you have settings enabled (also by default) which modify the file on save, like trim trailing whitespace, insert final newline, format etc etc?

**Comment by Friday32** [comment]
The problem was reproduced here on multiple systems. (5)

Today, I installed Windows 10 Pro on a virtual machine, installed Visual Studio Code 1.39.2, and ran the example without any VSCode setting changes.  The problem was reproduce-able.

**Comment by jrieken** [comment]
You are not answering my questions, the `files.trimFinalNewlines` is enabled by default, so if you have trailing newlines this will happen (on save, which happens when closing an editor since auto-save is used). 

Overall, this isn't a bug or a problem but expected behaviour, esp when a file is saved before/while being saved. Closing this as question until clear reproducible steps are provided that show unwanted/buggy behaviour.

**Comment by Friday32** [comment]
My response implies that I'm using default settings. ("files.trimFinalNewlines" is not checked by default).  But, clearly something in the default settings cause vscode to modify the document on close. I really don't care what it is. Even if the defaults were off, a user might switch them on.

Let me rephrase the problem.   I as the extension writer need to know whether or not the user made a change to their document. From your comments, I clearly cannot know this as VSCode may modify the document independently of user input.   From my original post "There is no way to determine why the method is called."

**Comment by jrieken** [comment]
That's something we don't know
